### PR TITLE
Make uploadstore package available outside of code intel

### PR DESCRIPTION
**Pull request by eseliger**
Might start using this in BC, so made this package publicly available.

### CODE REVIEW

File: enterprise/internal/codeintel/stores/lsifuploadstore/config.go [hunk 1]
```diff
@@ -10,7 +10,7 @@ import (
 )
 
 type Config struct {
-	*env.BaseConfig
+	env.BaseConfig
```
Comment (efritz) [inline enterprise/internal/codeintel/stores/lsifuploadstore/config.go:13]: this tripped me up

File: internal/uploadstore/config.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package uploadstore
+
+import (
+	"time"
+)
+
+type Config struct {
```
Comment (efritz) [inline internal/uploadstore/config.go:7]: Can we add a comment saying that this is _supplied in full_ by the user during construction of the upload store and **not** something that has a Load method that needs to be registered?

File: internal/uploadstore/store.go [hunk 3]
```diff
@@ -49,13 +48,21 @@ func CreateLazy(ctx context.Context, config *Config, observationContext *observa
...
 	}
 
-	store, err := newStore(ctx, config, newOperations(observationContext))
+	// Normalize the backend name.
+	config.Backend = strings.ToLower(config.Backend)
+
+	if config.Backend == "minio" {
+		// No manual provisioning
+		config.ManageBucket = true
+	}
```
Comment (efritz) [inline internal/uploadstore/store.go:63]: Could we extract this to a function in `config.go`? Something like `normalizeConfig`?

File: internal/uploadstore/config.go (hunk 2 above)
Comment (eseliger) [reply internal/uploadstore/config.go:7]: Would it help if it was called Options instead of Config?

Comment (efritz) [reply internal/uploadstore/config.go:7]: I think so yeah.

File: enterprise/cmd/precise-code-intel-worker/config.go [hunk 4]
```diff
@@ -5,24 +5,18 @@ import (
...
-	UploadStoreConfig  *uploadstore.Config
 	WorkerPollInterval time.Duration
 	WorkerConcurrency  int
 	WorkerBudget       int64
 }
 
 func (c *Config) Load() {
-	uploadStoreConfig := &uploadstore.Config{}
-	uploadStoreConfig.Load()
-	c.UploadStoreConfig = uploadStoreConfig
```
Comment (efritz) [inline enterprise/cmd/precise-code-intel-worker/config.go:24]: We should embed the new `lsifuploadstore` here instead, remove the new init function in that package, and back out the `init` changes to `enterprise/cmd/frontend/internal/codeintel/config.go`. If we embed it there as well everything works the same as before and we don't have to do any (additional) init tricks.

File: enterprise/cmd/frontend/internal/codeintel/services.go [hunk 5]
```diff
@@ -67,7 +63,7 @@ func NewServices(ctx context.Context, siteConfig conftypes.WatchableSiteConfig,
 	dbStore := store.NewWithDB(db, observationContext)
 	locker := locker.NewWithDB(db, "codeintel")
 	lsifStore := lsifstore.NewStore(codeIntelDB, siteConfig, observationContext)
-	uploadStore, err := uploadstore.CreateLazy(context.Background(), config.UploadStoreConfig, observationContext)
+	uploadStore, err := lsifuploadstore.New(context.Background(), observationContext)
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/services.go:66]: We can pass the config back through here again (can still validate it in the `New` method if you'd rather).

File: enterprise/cmd/frontend/main.go [hunk 6]
```diff
@@ -62,6 +62,12 @@ var initFunctions = map[string]EnterpriseInitializer{
 	"compute":        compute.Init,
 }
 
+var codeIntelConfig = &codeintel.Config{}
+
+func init() {
+	codeIntelConfig.Load()
+}
+
```
Comment (efritz) [inline enterprise/cmd/frontend/main.go:70]: These no longer need to be global if we embed the lsifuploadstore config.

Comment (eseliger) [reply enterprise/cmd/frontend/main.go:70]: Refactored this a bit and now we only need this singular place to load the config. To me this feels clean but LMK if you want this to look different.

**Comment by eseliger** [comment]
Main-dry-run: https://buildkite.com/sourcegraph/sourcegraph/builds/129580

File: internal/uploadstore/config.go (hunk 2 above)
Comment (efritz) [reply internal/uploadstore/config.go:10]: 🙉 📢  :)

Comment (eseliger) [reply internal/uploadstore/config.go:10]: Decided options wasn't very well fitting :D 
What does that comment mean? 👁️

File: enterprise/cmd/precise-code-intel-worker/config.go [hunk 7]
```diff
@@ -3,33 +3,30 @@ package main
...
-
 	c.WorkerPollInterval = c.GetInterval("PRECISE_CODE_INTEL_WORKER_POLL_INTERVAL", "1s", "Interval between queries to the upload queue.")
 	c.WorkerConcurrency = c.GetInt("PRECISE_CODE_INTEL_WORKER_CONCURRENCY", "1", "The maximum number of indexes that can be processed concurrently.")
 	c.WorkerBudget = int64(c.GetInt("PRECISE_CODE_INTEL_WORKER_BUDGET", "0", "The amount of compressed input data (in bytes) a worker can process concurrently. Zero acts as an infinite budget."))
+	c.LSIFUploadStoreConfig.Load()
```
Comment (efritz) [inline enterprise/cmd/precise-code-intel-worker/config.go:24]: This is being called on a value type.

File: enterprise/cmd/frontend/internal/codeintel/config.go [hunk 8]
```diff
@@ -2,37 +2,32 @@ package codeintel
...
-	uploadStoreConfig := &uploadstore.Config{}
-	uploadStoreConfig.Load()
-	config.UploadStoreConfig = uploadStoreConfig
-
+func (c *Config) Load() {
 	enqueuerConfig := &enqueuer.Config{}
 	enqueuerConfig.Load()
-	config.AutoIndexEnqueuerConfig = enqueuerConfig
+	c.AutoIndexEnqueuerConfig = enqueuerConfig
+	c.LSIFUploadStoreConfig.Load()
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/codeintel/config.go:22]: This is being called on a value type.

File: enterprise/cmd/precise-code-intel-worker/config.go (hunk 7 above)
Comment (eseliger) [reply enterprise/cmd/precise-code-intel-worker/config.go]: duh why is that even allowed when the receiver of Load is `*Config` 🙃 
da67073311a965db8f6369376869e3f88ed50299

File: enterprise/cmd/frontend/internal/codeintel/config.go (hunk 8 above)
Comment (eseliger) [reply enterprise/cmd/frontend/internal/codeintel/config.go:24]: da67073311a965db8f6369376869e3f88ed50299

**Review by efritz — APPROVED** [review]

**Comment by efritz** [comment]
Thanks! This was a pain 😅

**Comment by eseliger** [comment]
sorry for being a pain 😬

**Comment by eseliger** [comment]
Lol, forgot to merge.

**Pull request merged**